core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -0,0 +1,444 @@
+        IamMemberTypes.User,
+        organizationRoles,
+        condition = Some(condition)
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-user", tags = openTelemetryTags).unsafeToFuture()
+      _ <- googleIamDao.addRoles(
+        GoogleProject(googleProjectId.value),
+        userAndPet.petEmail,
+        IamMemberTypes.ServiceAccount,
+        organizationRoles,
+        condition = Some(condition)
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-pet", tags = openTelemetryTags).unsafeToFuture()
+    } yield ()
+  }
+
+  private def addUserAndPetToBucketIamRole(gcsBucketName: GcsBucketName,
+                                           organizationRoles: Set[String],
+                                           userAndPet: UserAndPetEmails,
+                                           condition: Expr,
+                                           googleProjectId: GoogleProjectId
+  ): Future[Unit] = {
+    logger.info(
+      s"Adding bucket-level FastPass access for $userAndPet in ${gcsBucketName.value} [${organizationRoles.mkString(" ")}]"
+    )
+    for {
+      _ <- googleStorageDAO.addIamRoles(
+        gcsBucketName,
+        userAndPet.userEmail,
+        IamMemberTypes.User,
+        organizationRoles,
+        condition = Some(condition),
+        userProject = Some(GoogleProject(googleProjectId.value))
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-user", tags = openTelemetryTags).unsafeToFuture()
+      _ <- googleStorageDAO.addIamRoles(
+        gcsBucketName,
+        userAndPet.petEmail,
+        IamMemberTypes.ServiceAccount,
+        organizationRoles,
+        condition = Some(condition),
+        userProject = Some(GoogleProject(googleProjectId.value))
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-pet", tags = openTelemetryTags).unsafeToFuture()
+    } yield ()
+  }
+
+  private def quotaAvailableForNewWorkspaceFastPassGrants(workspace: Workspace): Future[Boolean] =
+    for {
+      projectPolicy <- googleIamDao
+        .getProjectPolicy(GoogleProject(workspace.googleProjectId.value))
+        .map(fromProjectPolicy)
+      bucketPolicy <- googleStorageDAO
+        .getBucketPolicy(GcsBucketName(workspace.bucketName), Some(GoogleProject(workspace.googleProjectId.value)))
+        .map(fromBucketPolicy)
+    } yield {
+      // Role binding quotas do not de-duplicate member emails, hence the conversion of Sets to Lists
+      val numProjectRoleBindings = projectPolicy.bindings.toList.flatMap(_.members.toList).size
+      val expectedProjectBindings = numProjectRoleBindings + maxProjectRoleBindingsPerUser
+
+      val numBucketRoleBindings = bucketPolicy.bindings.toList.flatMap(_.members.toList).size
+      val expectedBucketBindings = numBucketRoleBindings + maxBucketRoleBindingsPerUser
+
+      expectedProjectBindings < policyBindingsQuotaLimit && expectedBucketBindings < policyBindingsQuotaLimit
```

Comment:
Do we also need to check that: 
```expectedProjectBindings + expectedBucketBinding < policyBindingsQuotaLimit``` ?  

Put another way, is the Policy Binding Quota at the Project level or at the resource (project/bucket) level?

Reply:
no, the policy binding limits are per resource type and projects vs buckets are different resource types